kernel/src/platform/scheduler_timer.rs

```diff
@@ -64,26 +64,6 @@ pub trait SchedulerTimer {
     /// layers are used.
     fn reset(&self);
 
-    /// Arm the SchedulerTimer timer and ensure an interrupt will be generated.
-    ///
-    /// The timer must already be started by calling `start()`. This function
-    /// guarantees that an interrupt will be generated when the already started
-    /// timer expires. This interrupt will preempt the running userspace
-    /// process.
-    fn arm(&self);
-
-    /// Disarm the SchedulerTimer timer indicating an interrupt is no longer
-    /// required.
-    ///
-    /// This does not stop the timer, but indicates to the SchedulerTimer that
-    /// an interrupt is no longer required (i.e. the process is no longer
-    /// executing). By not requiring an interrupt this may allow certain
-    /// implementations to be more efficient by removing the overhead of
-    /// handling the interrupt. The implementation may disable the underlying
-    /// interrupt if one has been set, depending on the requirements of the
-    /// implementation.
-    fn disarm(&self);
-
```

Comment:
I'm not sure why this interface has changed. We discussed this at length when adding the scheduler, and decided to decouple time keeping from interrupt generation. At the very least the comments in this file need to be updated.

However, I think changing this interface right after we just changed it should at least be its own PR.

Reply:
https://github.com/tock/tock/pull/2089#discussion_r481251980

Reply:
I actually think Brad is right -- I agreed too prematurely to the removal of `disarm()` and `arm()` from `sched.rs`. My mistake.

When we first added them to the interface, the idea was that they are an optimization, but are not required -- if a given `SchedulerTimer` implements both `arm()` and `disarm()`, it must implement bot, but if it implements neither, it can simply enable interrupts in `start()`. For all of the `SchedulerTimer` implementations that are not built on top of alarm, they can probably remain as they were, and for implementations built on top of alarm (aka `VirtualSchedulerTimer`), they can go without the optimization of disabling interrupts while not in userspace. So it probably makes more sense to keep `arm()` and `disarm()` in `sched.rs`, and not implement them for `VirtualSchedulerTimer`.

Reply:
OK, sounds good.

Reply:
This is a little more complicated. There isn't an `Alarm::enable` any more, in part due to concerns that you set the alarm to the future, disarm it, you pass the compare value, then re-arm it. Instead, you have to set the alarm again with set_alarm. Should I store those values in the struct and re-invoke `set_alarm`?

Reply:
This only applies to the `VirtualSchedulerTimer`, right? I am proposing this as the new implementation:

```rust
impl<A: 'static + time::Alarm<'static>> SchedulerTimer for VirtualSchedulerTimer<A> {
    fn reset(&self) {
        self.alarm.disable();
    }

    fn start(&self, us: u32) {
        let tics = {
            // We need to convert from microseconds to native tics, which could overflow in 32-bit
            // arithmetic. So we convert to 64-bit. 64-bit division is an expensive subroutine, but
            // if `us` is a power of 10 the compiler will simplify it with the 1_000_000 divisor
            // instead.
            let us = us as u64;
            let hertz = A::Frequency::frequency() as u64;

            (hertz * us / 1_000_000) as u32
        };
        let fire_at = self.alarm.now().wrapping_add(tics);
        self.alarm.set_alarm(fire_at);
    }

    fn arm(&self) {}

    fn disarm(&self) {}
```

Reply:
We should update the comments on the trait to indicate that `arm()`/`disarm()` are optional, but that if `disarm()` is implemented `arm()` must be implemented, and that if `arm()` is not implemented `start()` must enable interrupts.

Reply:
Happy to make a separate PR with those changes if that's easier.

Reply:
OK, sounds good.

Reply:
I will update the VirtualSchedulerTimer implementation accordingly.

Reply:
OK, fixed. `arm`/`disarm` are back in.